feat: enhance http service with request throttling logic based on all workers busy events

jorgeantonio21 opened the pull request:
#### Overview:

This PR implements a comprehensive request throttler system for the HTTP service that automatically activates when all KV router workers are busy. The system includes a configurable queue depth, NATS event-driven request throttling, and graceful degradation when request throttling is unavailable.

#### Details:

**Core Request Throttler Implementation:**
- Added HttpServiceRequestThrottler with configurable rejection time window
- Implemented RequestThrottlerState with automatic expiration logic
- Created NATS event subscription for "all workers busy" events from KV router scheduler
- Added request throttling checks to all HTTP endpoints (chat completions, completions, embeddings, responses)

**HTTP Service Integration:**
- Extended HttpServiceConfig with `all_workers_busy_rejection_time_window` parameter
- Updated State struct to include request throttler in shared service state
- Modified all request handlers to check request throttling before processing requests
- Added proper error responses (429 Too Many Requests) when request throttled

**Scheduler Queue Implementation:**
- Enhanced KV router scheduler with configurable queue depth (max_workers_busy_queue_depth)
- Added NATS event publishing when queue reaches maximum depth
- Implemented KVAllWorkersBusyEvent for communicating service status
- Added automatic queue clearing after configurable duration

**Service Lifecycle Management:**
- Integrated request throttler monitoring into HTTP service startup
- Added graceful error handling - service starts even if request throttling fails
- Proper separation of concerns between model discovery and request throttling
- Updated model watcher to remove request throttling responsibilities

**Configuration & Documentation:**
- Updated Python bindings to support new request throttling configuration
- Enhanced router README with queue depth documentation
- Added proper error types and logging throughout the system

#### Where should the reviewer start?

**Primary files to review:**
- lib/llm/src/http/service/request_throttler.rs - Core request throttling implementation
- lib/llm/src/http/service/service_v2.rs - HTTP service integration
- lib/llm/src/http/service/openai.rs - Request handler request throttling checks
- lib/llm/src/kv_router/scheduler.rs - Queue implementation and NATS events
- lib/llm/src/entrypoint/input/http.rs - Service startup and lifecycle management

**Secondary files:**
- lib/bindings/python/rust/llm/entrypoint.rs - Python binding updates
- docs/components/router/README.md - Documentation updates
- lib/llm/src/discovery/watcher.rs - Cleaned up model watcher

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)

- closes GitHub issue: #2165


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Added configurable queue depth for handling requests when all workers are busy, allowing users to set a maximum waiting queue size.
  * Introduced a configurable time window for rejecting new requests with a 429 error when all workers are busy.
  * Implemented event-driven request throttling, automatically rejecting requests during periods of high load based on system events.
  * Added Prometheus metrics to track request-throttled requests for improved monitoring and observability.

* **Documentation**
  * Updated documentation to explain new queue management and request throttling options, including configuration flags and event-driven behaviors.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jorgeantonio21 committed:
initial impl for all workers busy hook up with frontend rate limiter

jorgeantonio21 committed:
add workers busy args to entrypoint.rs bindings

jorgeantonio21 committed:
address ci failures

jorgeantonio21 committed:
address code rabbit PR comments

PeaBrane commented on lib/llm/src/kv_router/scheduler.rs:
```diff
@@ -136,70 +155,104 @@ impl KvScheduler {
+                            instances.iter().map(|i| i.instance_id).collect();
+                        pending_endpoint_update = Some(worker_ids);
+                    }
+                    Err(KvSchedulerError::AllWorkersBusy) => {
```
can you restore the TODO comment here that this is not hooked up. Currently, this error branch would never be executed because `select_worker` would not emit this error

PeaBrane commented on lib/llm/src/kv_router/scheduler.rs:
```diff
@@ -136,70 +155,104 @@ impl KvScheduler {
+                    is_waiting_request = true;
+                } else {
+                    // Waiting queue is empty, we can process new requests
+                    request = tokio::select! {
```
I think the nesting here needs to be inverted. In other words, `instances_rx.changed()` should always go through, as the Router needs to know immediately the new available endpoints for resource discovery.  So, the request queue logic should really be delegated to the `request_rx.recv()` arm only.

jorgeantonio21 replied:
You are right, this is actually a great observation

PeaBrane commented on lib/llm/src/kv_router/scheduler.rs:
```diff
@@ -117,15 +124,27 @@ impl KvScheduler {
             }
         });
 
+        let (all_workers_busy_event_tx, all_workers_busy_event_rx) =
```
I would make this channel bounded and introduce some sort of edge trigger mechanism. This event should only be sent when the threshold is exceeded from below, but should not be sent continuously when the queue is persistently above this threshold.

jorgeantonio21 committed:
address further PR comments

PeaBrane commented on lib/llm/src/http/service/rate_limiter.rs:
```diff
@@ -0,0 +1,161 @@
+}
+
+#[derive(Clone)]
+pub struct HttpServiceRateLimiter {
```
If we assume the timestamp is just the timestamp on event receive, then the events are naturally temporally ordered, and we can remove `clear_notify` and the inner nested tokio spawned process.

jorgeantonio21 committed:
Merge branch 'main' into feat/ja/all-workers-busy-rate-limiter

jorgeantonio21 committed:
address further PR comments

jorgeantonio21 committed:
reformat python e2e tests

jorgeantonio21 committed:
run pre-hooks

PeaBrane commented on tests/rate_limiter/test_rate_limiter_e2e_with_mockers.py:
```diff
@@ -0,0 +1,422 @@
+        logger.info("✓ Multiple requests consistently rate limited")
+
+        # Step 5: Wait for rate limiting to expire
+        logger.info("=== Waiting for rate limit to expire ===")
```
I would add one more second here to give some leeway to prevent flakiness

jorgeantonio21 committed:
add python e2e dependencies

jorgeantonio21 committed:
run pre-commit hooks

PeaBrane commented on lib/llm/src/http/service/rate_limiter.rs:
```diff
@@ -0,0 +1,353 @@
+    }
+
+    /// Check if state is currently rate limited
+    pub fn is_rate_limited(&self) -> bool {
+        if self.rate_limit_start.is_none() {
+            return false;
+        }
+
+        // Check if rate limit duration has expired
+        if let Some(start_time) = self.rate_limit_start {
+            if start_time.elapsed() > self.rate_limit_duration {
+                return false;
+            }
+        }
+        true
+    }
```
```suggestion
    pub fn is_rate_limited(&self) -> bool {
        match self.rate_limit_start {
            None => false,
            Some(start_time) => start_time.elapsed() <= self.rate_limit_duration
        }
    }
```

jorgeantonio21 committed:
address further PR comments

jorgeantonio21 committed:
Merge branch 'main' into feat/ja/all-workers-busy-rate-limiter

PeaBrane commented on lib/llm/src/entrypoint/input/http.rs:
```diff
@@ -28,10 +26,24 @@ pub async fn run(runtime: Runtime, engine_config: EngineConfig) -> anyhow::Resul
         EngineConfig::Dynamic(_) => {
             let distributed_runtime = DistributedRuntime::from_settings(runtime.clone()).await?;
+            let rate_limiter = http_service.rate_limiter_clone();
+            if let Err(e) = start_rate_limiter_monitoring(
```
Should we not start the rate limiter at all if we get None for `rejection_time_window`

PeaBrane commented on lib/llm/src/http/service/rate_limiter.rs:
```diff
@@ -0,0 +1,368 @@
+        namespace: Namespace,
+        cancellation_token: CancellationToken,
+    ) -> Result<()> {
+        let rate_limit_duration = { self.rate_limit_state.read().unwrap().rate_limit_duration };
```
we can trace a warning here if this is started with a none `time_window`

nnshah1 reviewed (commented):
Some general questions:

1) do we need a thread and a waiting time - instead could the router return a 'reject' on a request if the queue is full? We could avoid the synchronization here between the two - and perhaps simplify the logic. The router can return a "all workers busy" to the service. I understand there is slightly more processing involved - but I think the request flow simplifies a lot. We could also implement the timeout in that way - when we receive a "all workers busy rejection" we can reject all incoming requests until the timeout is over.

2) For my understanding - this now adds a queue to the router? Is that correct? but only in the kv router and not in round robin / etc.

3) if we separate the router out - I think router will no longer be able to have the queue - I think again this is where having the in the http side may make sense -and we can use the router only to deliver the "all busy" signal. The http portion can hold the queue and reject the request directly and schedule a timeout. 

3) Maybe we can rename this to "request throttling" or "Request Rejection" or "Backpressure" instead of "rate limiting" - I think that term brings to mind limiting the rate of requests over a period of time (like 5 requests per second per client) - and that's not the intent here IIUC. The goal here is to reject requests when the system is overloaded.

Awesome work - need some time to digest. 

@rmccorm4 , @kthui for viz as well.

jorgeantonio21 committed:
address further PR comments

jorgeantonio21 committed:
Merge branch 'main' into feat/ja/all-workers-busy-rate-limiter

jorgeantonio21 commented:
@PeaBrane I addressed your comments above, let me know if you have any other comments. Happy to address any further comments.

@nnshah1 we can definitely move the queue into the frontend, I see the value of having a single queued state for all routing mechanisms. 

Since this is open to discussion, let me know how you guys want to proceed

PeaBrane commented on lib/llm/src/http/service/rate_limiter.rs:
```diff
@@ -0,0 +1,136 @@
+    }
+
+    /// Check if the current underlying state is rate limited
+    pub fn is_rate_limited(&self) -> bool {
+        if !self.is_enabled {
+            return false;
+        }
+
+        let state = self.rate_limit_state.read().unwrap();
+        state.is_rate_limited()
+    }
```
fine for now

PeaBrane commented on lib/llm/src/http/service/rate_limiter.rs:
```diff
@@ -0,0 +1,136 @@
+                    let mut state = rate_limiter_clone.write().unwrap();
+                    state.clear_rate_limit();
+                });
+            }
```
stale

jorgeantonio21 committed:
address PR comments

jorgeantonio21 committed:
address further PR comments, merge main branch and resolve conflicts

jorgeantonio21 committed:
add request_thottler.rs file

jorgeantonio21 committed:
add e2e test and resolve merge issues

jorgeantonio21 committed:
address readme inconsistencies and further issues

jorgeantonio21 committed:
merge main and resolve conflicts

jorgeantonio21 committed:
refactor tests

jorgeantonio21 committed:
refactor e2e tests

jorgeantonio21 committed:
refactor e2e tests

jorgeantonio21 committed:
refactor e2e tests

jorgeantonio21 committed:
Merge branch 'main' into feat/ja/all-workers-busy-rate-limiter

jorgeantonio21 committed:
remove undefined args

jorgeantonio21 committed:
increase sleep duration

jorgeantonio21 committed:
enable setting endpoint-id for entrypoint args

jorgeantonio21 committed:
correct namespace parsing

jorgeantonio21 committed:
Merge branch 'main' into feat/ja/all-workers-busy-rate-limiter

jorgeantonio21 committed:
correct overflowing on sleep duration

jorgeantonio21 committed:
correct e2e tests

jorgeantonio21 committed:
run pre-commits

jorgeantonio21 committed:
Merge branch 'main' into feat/ja/all-workers-busy-rate-limiter

jorgeantonio21 committed:
merge main branch and resolve conflicts

jorgeantonio21 committed:
resolve merge issue
